base/common/src/org/dogtagpki/nss/NSSDatabase.java

```diff
@@ -343,13 +371,384 @@ public void addPEMCertificate(
         }
     }
 
+    public void addBasicConstraintsExtension(
```

Comment:
I guess I'm fine with this design for the time being but how would you envision transitioning this to JSS? One of your earlier `TODO`s mentioned you'd like to do that eventually.

Reply:
The other issue with this particular design is you aren't interpreting the output any. You're guessing whatever output `certutil` does _now_ and feeding it back as stdin to the program. What happens when `certutil` changes? I guess it'd be a lot more work to launch `certutil` interactively, feed input from config file / ... and see if it is already answered in the config file; if not, either prompt interactively or leave at its default value. 

IDK, seems potentially fragile. But perhaps we'll be able to rewrite with JSS later.

Reply:
The main goal is to provide a non-interactive tool to generate certs that can be used in ACME, `pkispawn`, `cert-fix`, etc. Ideally this functionality should have been provided by NSS tools, but as of now there seems to be no non-interactive way to generate certs using `certutil`. I've also looked at the JSS code in PKI for generating certs, but it's tied to the profile framework which may require some time to refactor.

Due to time constraints I had to pick the quicker solution, which is to use `certutil` for now. Hopefully we can replace this with JSS in the next release, or even in this release if there's still time.

It is indeed fragile, but so far NSS tools do not seem to change a lot. If it breaks we can always fix it, but I'd rather not rely on `certutil` exact output since that would add fragility even more. Note that the `NSSDatabase` Python class also has a similar code and it's still working fine. Eventually we need to change it to call `pki nss`.

The code basically consists of 2 parts: config parsing and `certutil` invocation. Since the config parsing already produces Java extension objects, later we only need to change the second part into JSS.

Reply:
I wasn't necessarily suggesting reusing PKI's profile-based JSS mechanism, but JSS provides raw PKCS#7 mechanisms for both CSRs and CRTs -- both under `org.mozilla.jss` and the `org.mozilla.jss.netscape.security` that aren't tied to PKI's profile mechanism.

Reply:
Sure, I'll check with you before we rewrite this code in JSS. I'm hoping to eventually have something that can work with both the new configuration file and the existing profile framework.

Reply:
We'll that's a big undertaking... Good luck! :D

Reply:
There are actually 2 profile frameworks: the bootstrap profile framework that is used to set up the server, and the main profile framework that runs after the server is up. I'm hoping the tool can use at least the bootstrap profiles (so it can be used in `pkispawn`), but will see about that later :)